### PR TITLE
Podtemplatespec

### PR DESCRIPTION
**Pull request by respringer**
Allow the user to override values in the init-container and three containers generated by cass-operator.  There are more details in the README.md.

### CODE REVIEW

File: operator/pkg/reconciliation/construct_service.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
+// Copyright DataStax, Inc.
```
Comment (respringer) [inline operator/pkg/reconciliation/construct_service.go:1]: There should be no changes to the logic in this file.  I simply moved service related functions to a new file.

File: operator/pkg/reconciliation/construct_statefulset.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
+// Copyright DataStax, Inc.
```
Comment (respringer) [inline operator/pkg/reconciliation/construct_statefulset.go:1]: Just like construct_service.go, there should be no logic changes in this file either.

File: operator/pkg/reconciliation/constructor.go [hunk 3]
```diff
@@ -6,351 +6,18 @@ package reconciliation
 // This file defines constructors for k8s objects
```
Comment (respringer) [inline operator/pkg/reconciliation/constructor.go:6]: This file should now contain miscellaneous construction logic for everything that does not have a dedicated construct_foo.go file.

File: README.md [hunk 4]
```diff
@@ -259,6 +259,57 @@ The operator is comprised of the following container images working in concert:
   4.0-beta1.
 * ... or DSE, built from [datastax/docker-images](https://github.com/datastax/docker-images).
 
+### Overriding properties of cass-operator created Containers
```
Comment (jimdickinson) [inline README.md:262]: excellent README addition!

File: operator/pkg/reconciliation/construct_podtemplatespec.go [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+					},
+				},
+			},
+		},
+	}
+}
+
+// calculatePodAntiAffinity provides a way to keep the db pods of a statefulset away from other db pods
+func calculatePodAntiAffinity(allowMultipleNodesPerWorker bool) *corev1.PodAntiAffinity {
+	if allowMultipleNodesPerWorker {
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/construct_podtemplatespec.go:53]: maybe we should hoist this arg and the if out of here and to the call site? would probably be more clear to read, right?

File: operator/pkg/reconciliation/construct_podtemplatespec.go [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+}
+
+func combineVolumeMountSlices(defaults []corev1.VolumeMount, overrides []corev1.VolumeMount) []corev1.VolumeMount {
+	out := append([]corev1.VolumeMount{}, overrides...)
+outerLoop:
+	// Only add the defaults that don't have an override
+	for _, volumeDefault := range defaults {
+		for _, volumeOverride := range overrides {
+			if volumeDefault.Name == volumeOverride.Name {
+				continue outerLoop
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/construct_podtemplatespec.go:124]: let's use a `found := false` boolean instead of the named continue

Comment (respringer) [reply operator/pkg/reconciliation/construct_podtemplatespec.go:124]: Ok.  I had that initially and it is more boilerplate-ish, but I can easily switch back.

File: operator/pkg/reconciliation/construct_podtemplatespec.go (hunk 5 above)
Comment (respringer) [reply operator/pkg/reconciliation/construct_podtemplatespec.go:53]: This is pre-existing code that I didn't change.  Can we update it in a new PR?

File: operator/pkg/reconciliation/construct_podtemplatespec.go [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+}
+
+func combineEnvSlices(defaults []corev1.EnvVar, overrides []corev1.EnvVar) []corev1.EnvVar {
+	out := append([]corev1.EnvVar{}, overrides...)
+outerLoop:
+	// Only add the defaults that don't have an override
+	for _, envDefault := range defaults {
+		for _, envOverride := range overrides {
+			if envDefault.Name == envOverride.Name {
+				continue outerLoop
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/construct_podtemplatespec.go:169]: same here and the other two places

Comment (jimdickinson) [reply operator/pkg/reconciliation/construct_podtemplatespec.go:169]: nvm

File: operator/pkg/reconciliation/construct_podtemplatespec.go (hunk 6 above)
Comment (jimdickinson) [reply operator/pkg/reconciliation/construct_podtemplatespec.go:124]: nvm

File: operator/pkg/reconciliation/reconcile_reaper.go [hunk 8]
```diff
@@ -15,52 +15,60 @@ import (
...
+	ReaperContainerName     = "reaper"
+	ReaperHealthCheckPath   = "/healthcheck"
+	ReaperKeyspace          = "reaper_db"
+	ReaperSchemaInitJob     = "ReaperSchemaInitJob"
 	// This code currently lives at https://github.com/jsanda/create_keyspace.
 	ReaperSchemaInitJobImage = "jsanda/reaper-init-keyspace:latest"
 )
 
-func buildReaperContainer(dc *api.CassandraDatacenter) corev1.Container {
+func buildReaperContainer(dc *api.CassandraDatacenter, container *corev1.Container) {
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/reconcile_reaper.go:35]: let's add a docstring to explain why we're modifying the input instead of returning something

Comment (jimdickinson) [reply operator/pkg/reconciliation/reconcile_reaper.go:35]: also let's add a note about adding a unit test here - but I know @jsanda has plans to change this code so no need to do it now

**Review by jimdickinson — APPROVED** [review]
just a couple small points - looks good to me!

**Comment by bradfordcp** [comment]
Legit awesome @respringer

**Pull request merged**